bgpd/bgp_labelpool.c

```diff
@@ -937,30 +1051,508 @@ DEFUN(show_bgp_labelpool_chunks, show_bgp_labelpool_chunks_cmd,
+	return 0;
+}
+
+static void labelpool_test_event_handler(struct thread *thread)
+{
+	struct lp_test *tcb;
+
+	if (skiplist_search(lp_tests, (void *)(uintptr_t)(lpt_generation),
+			    (void **)&tcb)) {
+
+		/* couldn't find current test in progress */
+		++lpt_test_event_tcb_lookup_fails;
+		return;
+	}
+
+	/*
+	 * request a bunch of labels
+	 */
+	for (unsigned int i = 0; (i < tcb->request_blocksize) &&
+				 (tcb->request_count < tcb->request_maximum);
+	     ++i) {
+
+		uintptr_t id;
+
+		++tcb->request_count;
+
+		/*
+		 * construct 32-bit id from request_count and generation
+		 */
+		id = ((uintptr_t)tcb->generation << 24) |
+		     (tcb->request_count & 0x00ffffff);
+		bgp_lp_get(LP_TYPE_VRF, (void *)id, test_cb);
+	}
+
+	if (tcb->request_count < tcb->request_maximum)
+		thread_add_event(bm->master, labelpool_test_event_handler, NULL,
+				 0, &tcb->event_thread);
+}
+
+static void lptest_stop(void)
+{
+	struct lp_test *tcb;
+
+	if (!lpt_inprogress)
+		return;
+
+	if (skiplist_search(lp_tests, (void *)(uintptr_t)(lpt_generation),
+			    (void **)&tcb)) {
+
+		/* couldn't find current test in progress */
+		++lpt_stop_tcb_lookup_fails;
+		return;
+	}
+
+	if (tcb->event_thread)
+		thread_cancel(&tcb->event_thread);
+
+	lpt_inprogress = false;
+}
+
+static int lptest_start(struct vty *vty)
+{
+	struct lp_test *tcb;
+
+	if (lpt_inprogress) {
+		vty_out(vty, "test already in progress\n");
+		return -1;
+	}
+
+	if (skiplist_count(lp_tests) >=
+	    (1 << (8 * sizeof(lpt_generation))) - 1) {
+		/*
+		 * Too many test runs
+		 */
+		vty_out(vty, "too many tests: clear first\n");
+		return -1;
+	}
+
+	/*
+	 * We pack the generation and request number into the labelid;
+	 * make sure they fit.
+	 */
+	unsigned int n1 = LPT_MAX_COUNT;
+	unsigned int sh = 0;
+	unsigned int label_bits;
+
+	label_bits = 8 * (sizeof(tcb->request_count) - sizeof(lpt_generation));
+
+	/* n1 should be same type as tcb->request_maximum */
+	assert(sizeof(n1) == sizeof(tcb->request_maximum));
+
+	while (n1 >>= 1)
+		++sh;
+	sh += 1; /* number of bits needed to hold LPT_MAX_COUNT */
+
+	if (sh > label_bits) {
+		vty_out(vty,
+			"Sorry, test iteration count too big on this platform (LPT_MAX_COUNT %u, need %u bits, but label_bits is only %u)\n",
+			LPT_MAX_COUNT, sh, label_bits);
+		return -1;
+	}
+
+	lpt_inprogress = true;
+	++lpt_generation;
+
+	tcb = XCALLOC(MTYPE_LABELPOOL_TEST, sizeof(*tcb));
+
+	tcb->generation = lpt_generation;
+	tcb->label_type = LP_TYPE_VRF;
+	tcb->request_maximum = LPT_MAX_COUNT;
+	tcb->request_blocksize = LPT_BLKSIZE;
+	tcb->labels = skiplist_new(0, NULL, NULL);
+	tcb->timestamps_alloc = skiplist_new(0, NULL, NULL);
+	tcb->timestamps_dealloc = skiplist_new(0, NULL, NULL);
+	thread_add_event(bm->master, labelpool_test_event_handler, NULL, 0,
+			 &tcb->event_thread);
+	monotime(&tcb->starttime);
+
+	skiplist_insert(lp_tests, (void *)(uintptr_t)tcb->generation, tcb);
+	return 0;
+}
+
+DEFUN(start_labelpool_perf_test, start_labelpool_perf_test_cmd,
```

Comment:
all new cli must be DEFPY http://docs.frrouting.org/projects/dev-guide/en/latest/workflow.html#pre-submission-checklist.  Should be easy to fix

Reply:
Will do